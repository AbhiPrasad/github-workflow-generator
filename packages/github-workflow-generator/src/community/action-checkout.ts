import type { CommunityStepOptions, StepOptions } from '../step';
import { UsesStep } from '../step';
import { dropUndefinedKeys } from '../utils';

const ACTION = 'actions/checkout' as const;

interface ActionCheckOutStepOptions {
  pin: string;
  // The branch, tag or SHA to checkout. When checking out the repository that
  // triggered a workflow, this defaults to the reference or SHA for that event.
  // Otherwise, uses the default branch.
  ref?: string;
  // Number of commits to fetch. 0 indicates all history for all branches and tags.
  // Default: 1
  fetchDepth?: number;
}

/**
 * Wrapper around the `actions/checkout` action to check out the current commit.
 *
 * https://github.com/actions/checkout
 */
export class ActionCheckoutStep extends UsesStep {
  constructor({ pin, ref, fetchDepth }: ActionCheckOutStepOptions = { pin: '' }, options: CommunityStepOptions = {}) {
    const actionName = pin.length ? `${ACTION}@${pin}` : ACTION;

    const stepOptions: StepOptions = {
      ...options,
    };

    const withOptions = dropUndefinedKeys({
      ref,
      'fetch-depth': fetchDepth,
    });

    if (Object.keys(withOptions).length) {
      stepOptions.with = withOptions;
    }

    super(actionName, stepOptions);
  }
}
