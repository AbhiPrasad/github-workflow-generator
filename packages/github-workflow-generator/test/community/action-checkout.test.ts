import { describe, expect, test } from 'vitest';
import { ActionCheckoutStep } from '../../src/community/action-checkout';

describe('ActionCheckoutStep', () => {
  test.each([
    ['regular', new ActionCheckoutStep()],
    ['with pinned version', new ActionCheckoutStep({ pin: 'v4' })],
    ['with pinned sha', new ActionCheckoutStep({ pin: 'dd126dd8c253650d181ad9538d8b4fa218fc31e8' })],
    ['with ref', new ActionCheckoutStep({ pin: 'v4', ref: '${{ env.HEAD_COMMIT }}' })],
    ['with fetchDepth', new ActionCheckoutStep({ pin: 'v4', fetchDepth: 2 })],
    ['with ref and fetchDepth', new ActionCheckoutStep({ pin: 'v4', ref: '${{ env.HEAD_COMMIT }}', fetchDepth: 2 })],
    ['with ref and fetchDepth', new ActionCheckoutStep({ pin: 'v4' }, { name: 'Checkout' })],
  ])('%s', (_, step) => {
    expect(step.toJSON()).toMatchSnapshot();
  });
});
