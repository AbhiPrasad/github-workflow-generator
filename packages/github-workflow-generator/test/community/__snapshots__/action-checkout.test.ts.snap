// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActionCheckoutStep > regular 1`] = `
{
  "uses": "actions/checkout",
}
`;

exports[`ActionCheckoutStep > with fetchDepth 1`] = `
{
  "uses": "actions/checkout@v4",
  "with": {
    "fetch-depth": 2,
  },
}
`;

exports[`ActionCheckoutStep > with pinned sha 1`] = `
{
  "uses": "actions/checkout@dd126dd8c253650d181ad9538d8b4fa218fc31e8",
}
`;

exports[`ActionCheckoutStep > with pinned version 1`] = `
{
  "uses": "actions/checkout@v4",
}
`;

exports[`ActionCheckoutStep > with ref 1`] = `
{
  "uses": "actions/checkout@v4",
  "with": {
    "ref": "\${{ env.HEAD_COMMIT }}",
  },
}
`;

exports[`ActionCheckoutStep > with ref and fetchDepth 1`] = `
{
  "uses": "actions/checkout@v4",
  "with": {
    "fetch-depth": 2,
    "ref": "\${{ env.HEAD_COMMIT }}",
  },
}
`;

exports[`ActionCheckoutStep > with ref and fetchDepth 2`] = `
{
  "name": "Checkout",
  "uses": "actions/checkout@v4",
}
`;
